patch-0002-Fix-race-condition-in-domain-suspension.patch

```diff
@@ -0,0 +1,108 @@
+From 9e7417c23e3eb51f0e17e8a1d994ebd1d387fb42 Mon Sep 17 00:00:00 2001
+From: Demi Marie Obenour <demi@invisiblethingslab.com>
+Date: Thu, 22 Sep 2022 12:27:32 -0400
+Subject: [PATCH 2/2] Fix race condition in domain suspension
+
+Check if the domain has suspended after setting the XenStore watch to
+prevent race conditions.  Also check if a guest has suspended when the
+timeout handler is called, and do not consider this to be a timeout.
+---
+ tools/libxl/libxl_dom_suspend.c | 27 +++++++++++++++++++++------
+ 1 file changed, 21 insertions(+), 6 deletions(-)
+
+diff --git a/tools/libxl/libxl_dom_suspend.c b/tools/libxl/libxl_dom_suspend.c
+index d80d495f4e16bbc29120125d47262af59aac74c2..b8cc94b652c02a779bfe46c17b561e478183af66 100644
+--- a/tools/libxl/libxl_dom_suspend.c
++++ b/tools/libxl/libxl_dom_suspend.c
+@@ -216,7 +216,7 @@ static void domain_suspend_common_wait_guest_evtchn(libxl__egc *egc,
+         libxl__ev_evtchn *evev);
+ static void suspend_common_wait_guest_watch(libxl__egc *egc,
+       libxl__ev_xswatch *xsw, const char *watch_path, const char *event_path);
+-static void suspend_common_wait_guest_check(libxl__egc *egc,
++static int suspend_common_wait_guest_check(libxl__egc *egc,
+         libxl__domain_suspend_state *dsps);
+ static void suspend_common_wait_guest_timeout(libxl__egc *egc,
+       libxl__ev_time *ev, const struct timeval *requested_abs, int rc);
+@@ -327,9 +327,11 @@ static void domain_suspend_common_pvcontrol_suspending(libxl__egc *egc,
+     STATE_AO_GC(dsps->ao);
+     xs_transaction_t t = 0;
+ 
+-    if (!rc && !domain_suspend_pvcontrol_acked(state))
++    if (!rc && !domain_suspend_pvcontrol_acked(state)) {
+         /* keep waiting */
++        LOGD(DEBUG, dsps->domid, "PV control callback without ack");
+         return;
++    }
+ 
+     libxl__xswait_stop(gc, &dsps->pvcontrol);
+ 
+@@ -402,6 +404,7 @@ static void domain_suspend_common_wait_guest(libxl__egc *egc,
+                                      60*1000);
+     if (rc) goto err;
+ 
++    suspend_common_wait_guest_check(egc, dsps);
```

Comment:
I don't think this explicit call is necessary. Xenstore watch is fired once immediately after being registered exactly to avoid this race.